projects/plugins/jetpack/modules/masterbar/admin-menu/class-dashboard-switcher-tracking.php

```diff
@@ -0,0 +1,167 @@
+<?php
+/**
+ * Quick switcher tracking file.
+ *
+ * @package automattic/jetpack
+ */
+
+namespace Automattic\Jetpack\Dashboard_Customizations;
+
+use Automattic\Jetpack\Terms_Of_Service;
+use Automattic\Jetpack\Tracking;
+use Jetpack_Plan;
+
+/**
+ * Class Dashboard_Switcher_Tracking
+ */
+class Dashboard_Switcher_Tracking {
+
+	/**
+	 * Jetpack Tracking library will prefix the event name with "jetpack_*" automatically.
+	 */
+	const JETPACK_EVENT_NAME = 'dashboard_quick_switch_link_clicked';
+
+	const WPCOM_EVENT_NAME = 'wpcom_dashboard_quick_switch_link_clicked';
+	/**
+	 * Jetpack tracking object.
+	 *
+	 * @var Tracking
+	 */
+	private $tracking;
+
+	/**
+	 * Current site plan.
+	 *
+	 * @var string
+	 */
+	private $plan;
+
+	/**
+	 * The wpcom_tracks wrapper function.
+	 *
+	 * @var callable
+	 */
+	private $wpcom_tracking;
+
+	/**
+	 * Dashboard_Switcher_Tracking constructor.
+	 *
+	 * @param Tracking $tracking       Jetpack tracking object.
+	 * @param callable $wpcom_tracking A wrapper over wpcom event record.
+	 * @param string   $plan           The current site plan.
+	 */
+	public function __construct( Tracking $tracking, callable $wpcom_tracking, $plan ) {
+		$this->tracking       = $tracking;
+		$this->plan           = $plan;
+		$this->wpcom_tracking = $wpcom_tracking;
+	}
+
+	/**
+	 * Create an event for the Quick switcher when the user changes it's preferred view.
+	 *
+	 * @param string $screen The screen page.
+	 * @param string $view   The new preferred view.
+	 */
+	public function record_switch_event( $screen, $view ) {
+		$event_props = array(
+			'current_page' => $screen,
+			'destination'  => $view,
+			'plan'         => $this->plan,
+		);
+
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$event_props['blog_id'] = get_current_blog_id();
+			$wpcom_tracking         = $this->wpcom_tracking;
+			$wpcom_tracking( $event_props );
```

Comment:
Nitpick: Can this be changed to `$this->record_wpcom_event( $event_props )` and create a new `record_wpcom_event` function which does something like below?

```php
private function record_wpcom_event( $event_props ) {
	$this->wpcom_tracking( $event_props );
}
```

We can also remove the `wpcom_tracks_record_event` function and include its implementation in this new `record_wpcom_event` function:

```php
private function record_wpcom_event( $event_props ) {
	jetpack_require_lib( 'tracks/client' );
	\tracks_record_event( \wp_get_current_user(), self::WPCOM_EVENT_NAME, $event_props );
}
```

That way I think the code is more legible because both the WPCOM and the Jetpack flows are aligned. Otherwise, I found a bit hard to understand the code because of the amount of functions I need to remember while mentally parsing the code.

Reply:
The idea was to isolate the external dependencies outside of the implementation itself so that we can:
- Have a public API of a class with just one method: `record_switch_event`
- Helper methods that can be easily extracted outside of the implementation. One example of this is ::get_plan which could be reused in the admin-menu. In this case, we would simply move the method outside and update the parameter when we create an instance of the class. In this manner, we will not have to make any internal changes in the switcher tracking class. 

It does indeed add a bit more complexity for wpcom_tracking callable but I thought makes it consistent with how we inject the Jetpack Tracker object. Besides, we can test the implementation easier this way just by using plain PHP.